sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,368 @@
+
+This extension is written against the Level Zero v1.5.8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+== API
+
+* Enumerations
++
+```
+ze_bindless_images_ext_version_t
+```
+* Structures
++
+```
+ze_image_mem_exp_desc_t
+ze_bindless_image_exp_desc_t
+ze_image_mem_alloc_exp_desc_t
+```
+
+== Overview
+
+Bindless images are a feature that provides flexibility on how images are accessed and used, such as removing limitations on how many images can be accessed as well as potentially improving performance.
+This is an improvement on the legacy of binding images which is a holdover from binding slots in hardware which used to be limited in number.
+
+Currently, bindless images are used internally in Level Zero.
+However, the full capabilities of bindless images are not exposed such as being able to independently access different types of images in an array regardless of their dimensionality as long as the images are accessed correctly by the user.
+
+This document will be proposing an extension for Level Zero to enable bindless images to be fully exploited.
+Part of the inspiration for this proposal is to allow the implementation of bindless images in SYCL, a proposal for which is currently in development, where Level Zero is used as a backend.
+
+
+=== Current Status of Images in Level Zero
+
+Currently, `zeImageCreate` does not produce image handles that have the properties to enable bindless images.
+In addition to this, images allocate and own their memory.
+These two things conflict with work being conducted on SYCL bindless images.
+This is because the API for SYCL bindless images has been designed to improve the flexibility and general usability of images in SYCL and this introduces some conflicts with Level Zero.
+For bindless images in SYCL to work in Level Zero, not only must there be a way to produce bindless image handles, but also allocate and populate image memory before `zeImageCreate` is called.
+
+Level Zero stores image memory in an implementation-specific encoding and layout that optimizes for device access such as the z-order curve.
+This proposal, instead, focuses on storing image memory in a linear fashion.
+
+Also, accessing images through USM is not supported by Level Zero.
+This would allow the reading and writing of image data with conventional pointers, in contrast to the current Level Zero images which require the exclusive use of API.
+This proposal introduces bindless images to Level Zero which allows for USM access.
```

Comment:
We you need to access image via USM pointer?

Reply:
A requirement for our extension was to allow for the creation of bindless images via USM. A potential use-case is re-using existing USM memory preventing copies, creating an image on top of that memory, and then accessing the image with hardware sampling applied to the data.

Reply:
Do you have plans to access such image from the host side via such USM pointer, like reading pixels.
Or is it only to provide backing storage for such image?
What are the use-cases for this extension?

Reply:
As of right now, we have only worked with USM bindless images on a device resident basis. We propose that images may be created from linear memory in this extension. That way, dereferencing the pointer would return the expected image data.

I would imagine that the performance would be dependent on the use-case. Allocating new memory for a tiled layout and copying the linear layout image data to this new allocation could be slower than simply allowing for the creation of an image on top of the already existing linear data and sampling that (despite the slower sampling speed).

I don't think we've seen the L0 Image View extension. Will take a look into this.